### PR TITLE
feat: add schema and validate-pyproject support

### PR DESCRIPTION
**Pull request by henryiii**
Adding a json schema. This supports validate-pyproject, allowing it to check a `tool.cibuildwheel` section. I wrote a Python 3.10+ `bin/generate_schema.py` to make the schema as I didn't see a good way to do the OS sections and overrides by hand. I mostly tried to replicate the current behavior, though we technically could be more strict with a schema than in reality, since it's optional to validate it.

Testing by validating the defaults toml, a few manual examples, and running all the docs examples though.

Once we have this, we could also contribute it to SchemaStore, which would make it work in vscode, etc.

### CODE REVIEW

File: bin/generate_schema.py [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+
+overrides["items"]["properties"]["select"]["oneOf"] = string_array
+overrides["items"]["properties"] |= limited.copy()
+
+del limited["archs"]
+
+not_linux = limited.copy()
+
+del not_linux["environment-pass"]
+del not_linux["container-engine"]
```
Comment (joerick) [inline bin/generate_schema.py:216]: container-engine is actually only acceptable as a global option

File: bin/generate_schema.py [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+items:
+  type: object
+  required: ["select"]
+  additionalProperties: false
+  properties:
+    select: {}
+"""
+)
+
+limited = schema["properties"].copy()
```
Comment (joerick) [inline bin/generate_schema.py:204]: Perhaps this variable would be more clear named as `non_global_options`?

File: bin/generate_schema.py [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+del limited["build"]
+del limited["skip"]
+
```
Comment (joerick) [inline bin/generate_schema.py:207]: Also, `test-skip` and `archs` are global. Archs cannot be overridden in an override.

File: noxfile.py [hunk 4]
```diff
@@ -113,6 +113,14 @@ def update_proj(session: nox.Session) -> None:
     )
 
 
+@nox.session(reuse_venv=True)
+def generate_schema(session: nox.Session) -> None:
+    session.install("pyyaml", "ruamel.yaml")
```
Comment (joerick) [inline noxfile.py:118]: I think that ruamel.yaml isn't used?
```suggestion
    session.install("pyyaml")
```

**Review by joerick — COMMENTED** [review]
This is really cool, I didn't know about this. Getting auto-complete in vscode would be amazing.

**Comment by henryiii** [comment]
Also added a mechanism to test all the docs examples. Several fixes of the schema came from that. :)

File: docs/options.md [hunk 5]
```diff
@@ -917,9 +917,7 @@ Platform-specific environment variables are also available:<br/>
     CIBW_REPAIR_WHEEL_COMMAND: >
       python scripts/repair_wheel.py -w {dest_dir} {wheel} &&
       python scripts/check_repaired_wheel.py -w {dest_dir} {wheel}
-    ```
```
Comment (henryiii) [inline docs/options.md:920]: FYI, I could handle this, I suppose, but it's easier to just leave these as one block, like all the other examples.

**Review by joerick — APPROVED** [review]
Looks good! Testing with the docs samples is a very cool idea.

**Pull request merged**